File: class3/oop/oop-py/shapes.py
from abc import ABC, abstractmethod

class Shape(ABC):
    def __init__(self, base, altura): # Llaman esto
        super().__init__()
        self.base = base
        self.altura = altura
        self.tipo = "Shape"
    
    @abstractmethod
    def area(self):
        pass

    def description(self):
        return f"{self.tipo} de base {self.base} y altura {self.altura}"
    


class Triangle(Shape):

    def area(self):
        return self.base * self.altura/2
    
    def description(self):
        return f"un triangulo muy bonito de base {self.base} y altura {self.altura}"
    
    def color(self):
        return "azul"
    
    
class Rectangle(Shape):
    def __init__(self, base, altura):
        super().__init__(base, altura) #Esto
        self.tipo = "Rectangulo"
    
    def area(self):
        return self.base * self.altura
    
   

File: class3/oop/oop-rs/src/shapes.rs
pub trait Shape {
    fn area(&self) -> f64;
    fn report(&self);
}

pub trait Color {
    fn color(&self) -> &'static str {
        "red"
    }
}

impl Color for Triangle {}

impl Color for Rectangle {
    fn color(&self) -> &'static str {
        "blue"
    }
}

pub struct Triangle {
    base: f64,
    height: f64,
}

pub struct Rectangle {
    width: f64,
    height: f64,
}

impl Triangle {
    pub fn new(base: f64, height: f64) -> Triangle {
        Triangle { base, height }
    }
    pub fn color(&self) -> &'static str {
        "red"
    }
}

impl Rectangle {
    pub fn new(width: f64, height: f64) -> Rectangle {
        Rectangle { width, height }
    }
}

impl Shape for Triangle {
    fn area(&self) -> f64 {
        0.5 * self.base * self.height
    }

    fn report(&self) {
        println!(
            "I am a triangle with base {} and height {}",
            self.base, self.height
        );
    }
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }

    fn report(&self) {
        println!(
            "I am a rectangle with width {} and height {}",
            self.width, self.height
        );
    }
}
